### PR TITLE
Tests and performance fixes for hash helpers.

### PR DESCRIPTION
**Pull request by stebet**
This PR has the following changes:
* Improved performance and reduced allocations for SHA1 hash helpers 🚀
  * `IsStringSHA1Hash` is now hundreds of times faster and doesn't allocate any memory except for the empty string case which is almost a nanosecond slower now 😱
  * `CreateSHA1Hash` is now around 3 times faster and allocates 4 times less than previously. It can also be improved even more once the project moves to .NET Core so stay tuned :)
* SHA1 hash helpers are now extensions methods
* Added unit tests for SHA1 hash helpers

```ini
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.21390
Intel Core i7-10700 CPU 2.90GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4395.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4395.0), X64 RyuJIT
```

|              Type | Method |           Input |         Mean |      Error |     StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |------- |---------------- |-------------:|-----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|  IsStringSHA1Hash | Before |                 |     2.091 ns |  0.0243 ns |  0.0215 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|  IsStringSHA1Hash |  After |                 |     2.872 ns |  0.0205 ns |  0.0182 ns |  1.37 |    0.02 |      - |     - |     - |         - |
|  IsStringSHA1Hash | Before | SCOTTHELMESUCKS |   918.974 ns |  5.2620 ns |  4.3940 ns | 1.000 |    0.00 | 0.0553 |     - |     - |     353 B |
|  IsStringSHA1Hash |  After | SCOTTHELMESUCKS |     3.807 ns |  0.0334 ns |  0.0313 ns | 0.004 |    0.00 |      - |     - |     - |         - |
|  IsStringSHA1Hash | Before |           F4A69 |   940.215 ns | 10.0856 ns |  9.4341 ns | 1.000 |    0.00 | 0.0553 |     - |     - |     353 B |
|  IsStringSHA1Hash |  After |           F4A69 |     3.791 ns |  0.0349 ns |  0.0326 ns | 0.004 |    0.00 |      - |     - |     - |         - |
|                   |        |                 |              |            |            |       |         |        |       |       |           |
|    CreateSHA1Hash | Before |       Passw0rd1 | 1,618.698 ns | 13.5409 ns | 12.0037 ns |  1.00 |    0.00 | 0.2842 |     - |     - |   1,797 B |
|    CreateSHA1Hash |  After |       Passw0rd1 |   532.415 ns |  3.1252 ns |  2.6097 ns |  0.33 |    0.00 | 0.0677 |     - |     - |     433 B |
|    CreateSHA1Hash | Before |         hunter2 | 1,617.232 ns | 14.0797 ns | 13.1701 ns |  1.00 |    0.00 | 0.2842 |     - |     - |   1,797 B |
|    CreateSHA1Hash |  After |         hunter2 |   534.769 ns |  7.9382 ns |  7.4254 ns |  0.33 |    0.01 | 0.0677 |     - |     - |     433 B |

### CODE REVIEW

**Comment by mdawsonuk** [comment]
@stebet Awesome work! I was wondering who would add tests first 😁 Hopefully this is merged soon so I can update the NTLM hash function on my PR

File: Functions/Hash.cs [hunk 1]
```diff
@@ -9,48 +9,69 @@ namespace Functions
...
+            {
+                hexChars[index++] = HexChars[(hash[i] >> 4) & 0b1111];
+                hexChars[index++] = HexChars[(hash[i]) & 0b1111];
             }
+
+            return new string(hexChars);
         }
 
         /// <summary>
         /// Check that the string is a valid SHA-1 hash with regex
```
Comment (BBosman) [inline Functions/Hash.cs:50]: I think this comment is now out of date, as it no longer uses a `regex`.

File: Functions/Hash.cs [hunk 2]
```diff
@@ -9,48 +9,69 @@ namespace Functions
...
         /// </summary>
         /// <param name="input">Input hash to check</param>
         /// <returns>Boolean representing if the input is valid or not</returns>
-        public static bool IsStringSHA1Hash(string input)
+        public static bool IsStringSHA1Hash(this string input) => input.IsHexStringOfLength(40);
+        
+        public static bool IsHexStringOfLength(this string input, int requiredLength)
```
Comment (BBosman) [inline Functions/Hash.cs:56]: Doesn't this need documentation? Like:
```cs
/// <summary>
/// Check that the string contains only hexadecimal characters and is of a given length.
/// </summary>
/// <param name="input">Input string to verify</param>
/// <param name="requiredLength">The length the input should be</param>
/// <returns>Boolean representing if the input icontains only hexadecimal charactars and is the expected length</returns>
```

Comment (stebet) [reply Functions/Hash.cs:56]: Correct. I'll add it :)

File: Functions/Hash.cs (hunk 1 above)
Comment (stebet) [reply Functions/Hash.cs:50]: Nice catch.

**Pull request merged**